### ISSUE TITLE
How could I get the best hyperparameter when there is high stochasticity?

### ISSUE THREAD

**Issue by junqiqian** [optuna/optuna#2600]
The set of best hyperparameters I get is found to be unstable or "lucky". It is hard to be reproduced. The performance of the reproduction experiment with the same hyperparameters is much lower than the value as before. 
Now I use one trial for one set of hyperparameters. 

**Is it possible to avoid the previous issue without adding more trials for each set of hyperparameters (otherwise it would be expensive)?**
**Is there a way to have access to the mode of distribution of the TPE sampler so that I can know what is the best expected hyperparameters?**

**Comment by HideakiImamura** [comment]
Thanks for the question. I have had a similar problem before and have considered ways to deal with it. I don't know if this will be helpful but please have a look.

> Is it possible to avoid the previous issue without adding more trials for each set of hyperparameters (otherwise it would be expensive)?

In principle, I don't think it is possible to reduce the noise on the objective function value of each hyperparameter without adding more trials. It would be difficult to put any effort into the optimization itself to deal with this problem. Instead, we can put some effort into analyzing the results after optimization as follows.

> Is there a way to have access to the mode of distribution of the TPE sampler so that I can know what is the best expected hyperparameters?

No, I think we cannot calculate the posterior mean of the distribution of TPE. In TPE, we model the probability `p(x | y < c, D_n)` where `x` is a hyperparameter, `y` is the objective value, `c` is a constant and `D_n` is data collected until `n` steps. We cannot calculate the posterior mean, which is `E[y | x, D_n]` for given `x`, in principle.

**Comment by fsmosca** [comment]
> The set of best hyperparameters I get is found to be unstable or "lucky". It is hard to be reproduced. The performance of the reproduction experiment with the same hyperparameters is much lower than the value as before.
> Now I use one trial for one set of hyperparameters.

Could you describe you objective function?

**Comment by junqiqian** [comment]
> Thanks for the question. I have had a similar problem before and have considered ways to deal with it. I don't know if this will be helpful but please have a look.
> 
> > Is it possible to avoid the previous issue without adding more trials for each set of hyperparameters (otherwise it would be expensive)?
> 
> In principle, I don't think it is possible to reduce the noise on the objective function value of each hyperparameter without adding more trials. It would be difficult to put any effort into the optimization itself to deal with this problem. Instead, we can put some effort into analyzing the results after optimization as follows.
> 
> > Is there a way to have access to the mode of distribution of the TPE sampler so that I can know what is the best expected hyperparameters?
> 
> No, I think we cannot calculate the posterior mean of the distribution of TPE. In TPE, we model the probability `p(x | y < c, D_n)` where `x` is a hyperparameter, `y` is the objective value, `c` is a constant and `D_n` is data collected until `n` steps. We cannot calculate the posterior mean, which is `E[y | x, D_n]` for given `x`, in principle.
> 
> As a result, you cannot calculate the posterior mean with TPE. However, you can calculate the posterior mean with some surrogate models (such as Gaussian processes) after optimization as the following pseudo code.
> 
> ```python
> # Do optimization with TPESampler.
> study = optuna.create_study()
> study.optimize(objective, n_trials=...)
> 
> # Create a surrogate model.
> model = CreateGP(study)
> 
> # Calculate posterior means.
> means = [model.calculate_poseterior_mean(trial) for trial in study.trials]
> 
> # Get best trial along above posterior means
> best_trial = study.trials[numpy.argmin(means)]
> ```

Thanks for your answer!

Sorry I did not clarify the second question. I am actually interested in the expected hyperparameters of TPE sampler, which is `E[x | y < c, D_n]`. Is there any way to have access to this value?

**Comment by junqiqian** [comment]
> Could you describe you objective function?

The objective of my study is the output value of neural networks.

**Comment by fsmosca** [comment]
> > Could you describe you objective function?
> 
> The objective of my study is the output value of neural networks.

So you have outputs net_0, net_1 ... net_n where n >= 1 and then evaluate each of them. net_0 performed 80% (high is better), net_1 is only 75%, and it turns out the best net is net_0 from hyperparameters a=1, b=2 for example. Now when you try again the hyperparameters a=1, b=2 with resulting net_0x, this net_0x did not perform well compared to net_0. So perhaps try to look for the random seed used in this whole process. Fix the random seed and see if you can get repeatability of performance.

The other way is to change the objective, use the performance of the net as your objective. Before doing so measure the noise of this performance. Then during trials when sending this performance to optimizer add the noise - it will act as a margin.

**Comment by bouthilx** [comment]
Executing HPO once on a single seed is hazardous. As you saw the results are not reliable and could be very different for different seeds. 

@fsmosca I don't see how optimizing `f(x) - constant` would be different than optimizing `f(x)` directly. Unless you mean estimating variance of f(x) using a binomial approximation of the accuracy? In this case it is not a constant indeed, but it only attempts to model variance due to data sampling and ignore all other sources of variance. 

The best solution is executing n trainings for a given set of hyperparameters and return the average to the HPO algorithm. But it is quite expensive...

To compute the expected objective based on hyperparameters, the best approach would be the one suggested by @HideakiImamura. You need to be careful thought as there is no guarantee that the Gaussian process is reliable. First reason is that you need to make sure you covered the search space well enough for the GP to be reliable and this is hard to know. Second reason is that many search spaces have sharp edges which cause small modeling errors to have a large impact. You can see Appendix C of this paper where they discuss this as the 'danger zone': https://arxiv.org/abs/2007.01547. 

If your goal is to compare ML algorithms the best cheap option is to compare them using a statistical test so that you can estimate the average performance and uncertainty of this estimation. We explain this in our recent work at MLSys: https://www.youtube.com/watch?v=-bukIBTkQiM. This way you can avoid running multiple trials for each set of hyperparameters but still make a comparison between ML algos that is robust to noise.

**Comment by HideakiImamura** [comment]
Thanks all for the fruitful discussions!

@junqiqian 
>  I am actually interested in the expected hyperparameters of TPE sampler, which is E[x | y < c, D_n]. Is there any way to have access to this value?

No, we don't have any public interface to access the value in Optuna. But, we can calculate it in principle by directly using the Parzen estimator of TPE. Do you have any specific reason to use `E[x | y < c, D_n]`? If it is reasonable, we (core-dev members) will develop the public method to calculate `E[x | y < c, D_n]` (or create a contribution-welcome issue).

**Comment by fsmosca** [comment]
@bouthilx 

> Executing HPO once on a single seed is hazardous. As you saw the results are not reliable and could be very different for different seeds. 

The idea is to check where the noise is coming from.

> 
> @fsmosca I don't see how optimizing `f(x) - constant` would be different than optimizing `f(x)` directly. Unless you mean estimating variance of f(x) using a binomial approximation of the accuracy? In this case it is not a constant indeed, but it only attempts to model variance due to data sampling and ignore all other sources of variance.
> 
> The best solution is executing n trainings for a given set of hyperparameters and return the average to the HPO algorithm. But it is quite expensive...

That is not `constant` it has to be calculated. When doing trials, it is generally not good to just send an objective value that is not telling the truth to the optimizer, the noise has to be taken into account before sending it to optimizer. Even the optimizer will also get confused which may lead to long trials whose output is unreliable.

**Comment by bouthilx** [comment]
> That is not constant it has to be calculated. When doing trials, it is generally not good to just send an objective value that is not telling the truth to the optimizer, the noise has to be taken into account before sending it to optimizer. Even the optimizer will also get confused which may lead to long trials whose output is unreliable.

Indeed, but calculating this noise typically requires multiple trainings as most sources of variance are coming from the training phase rather than from the evaluation phases (computing validation accuracy). Running HPO once and then computing variance of multiple trainings can be a middle ground as we explain here https://www.youtube.com/watch?v=-bukIBTkQiM.

**Comment by junqiqian** [comment]
Thanks for your reply!

> Do you have any specific reason to use `E[x | y < c, D_n]`? If it is reasonable, we (core-dev members) will develop the public method to calculate `E[x | y < c, D_n]` (or create a contribution-welcome issue).

Yes, using E[x | y<c, D_n] would have more confidence than using the sampled x directly if we can choose the appropriate c. The expected value should be more reliable than a single value since it could avoid the violent perturbation caused by high stochasticity to some extent. It would be helpful if we can have access to this expected value.

**Comment by Turakar** [comment]
I share this problem of high stochasticity in training. To my problem, only a small dataset is available, and although I already use 10-fold cross-validation for estimating the performance of my CNN, I can clearly observe noisy trial values which lead to bad parameter choices if one just naively takes the minimum. For example, consider the excerpt from the slice plot attached to this post:

![slice plot with a noisy trial](https://user-images.githubusercontent.com/3958680/120110805-64b44b00-c16f-11eb-9fd8-efc02da04a89.png)

One can observe that it seems to be best to use no hidden layer at all. But this single trial with `dense_hidden = 1` has a low value induced by a high noise, although all other trials with `dense_hidden = 1` are clearly worse. With just one dimension, this is easy to observe by the human looking at the slice plots, but for many covariant parameters (e.g. hidden layers and dropout) this quickly becomes infeasible. I think that a unified modelling of the trial value space would be helpful here. I personally tried GP and SVMs, but this has shown no improvement.

**Comment by Turakar** [comment]
As a follow up to my previous post, I can say that I found a solution for my case:

1. Repeat the HPO multiple times &rarr; due to randomness the chosen HPO values are different
2. Of each study, select the best trial just by its value (in my case, the trial value is determined by repeated holdout validation)
3. Create an ensemble of the models determined by each study's best trial.

The ensemble is just a mean in my case, as I have a regression task. For classification, one might consider majority voting or adding the probabilities or something along these lines.
